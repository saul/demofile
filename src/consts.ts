export const MAX_EDICT_BITS = 11;
export const MAX_EDICTS = 1 << MAX_EDICT_BITS;
export const NETWORKED_EHANDLE_ENT_ENTRY_MASK = MAX_EDICTS - 1;
export const NUM_NETWORKED_EHANDLE_SERIAL_NUMBER_BITS = 10;
export const NUM_NETWORKED_EHANDLE_BITS =
  MAX_EDICT_BITS + NUM_NETWORKED_EHANDLE_SERIAL_NUMBER_BITS;
export const MAX_OSPATH = 260;
export const MAX_SPLITSCREEN_CLIENTS = 260;
export const MAX_PLAYER_NAME_LENGTH = 128;
export const SIGNED_GUID_LEN = 32;
export const MAX_CUSTOM_FILES = 4;
export const SUBSTRING_BITS = 5;
export const MAX_USERDATA_BITS = 14;
export const INVALID_NETWORKED_EHANDLE_VALUE =
  (1 << NUM_NETWORKED_EHANDLE_BITS) - 1;
