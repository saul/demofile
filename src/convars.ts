import { EventEmitter } from "events";
import { DemoFile } from "./demo";
import { CNETMsgSetConVar } from "./protobufs/netmessages";

interface IConVarChangeEvent {
  name: string;
  value: string;
  oldValue?: string;
}

export declare interface ConVars {
  /**
   * Fired when any console variable is changed (e.g., 'mp_buytime').
   */
  on(
    event: "change" | string,
    listener: (event: IConVarChangeEvent) => void
  ): this;
  emit(name: "change" | string, event: IConVarChangeEvent): boolean;
}

/**
 * Manages console variables.
 */
export class ConVars extends EventEmitter {
  public vars: Map<string, string> = new Map();

  public listen(demo: DemoFile): void {
    demo.on("net_SetConVar", (msg: CNETMsgSetConVar) => {
      const convars = msg.convars;
      if (!convars) return;
      for (const cvar of convars.cvars) {
        const oldValue = this.vars.get(cvar.name);
        this.vars.set(cvar.name, cvar.value);

        const args = {
          name: cvar.name,
          value: cvar.value,
          oldValue
        };

        this.emit(cvar.name, args);
        this.emit("change", args);
      }
    });
  }
}
