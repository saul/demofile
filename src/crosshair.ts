import { decodeShareCode } from "./sharecode";

export enum CrosshairStyle {
  Default = 0,
  DefaultStatic,
  Classic,
  ClassicDynamic,
  ClassicStatic
}

export interface ICrosshairInfo {
  style: keyof typeof CrosshairStyle;
  hasCenterDot: boolean;

  length: number;
  thickness: number;
  gap: number;

  hasOutline: boolean;
  outline: number;

  red: number;
  green: number;
  blue: number;

  hasAlpha: boolean;
  alpha: number;

  splitDistance: number;
  innerSplitAlpha: number;
  outerSplitAlpha: number;
  splitSizeRatio: number;

  isTStyle: boolean;
}

export function decodeCrosshairCode(shareCode: string): ICrosshairInfo {
  function signed(byte: number): number {
    return byte > 127 ? -(~byte & 0xff) - 1 : byte;
  }

  const bytes = decodeShareCode(shareCode);

  if (bytes[1] !== bytes.slice(2).reduce((x, y) => x + y) % 256) {
    throw new Error("invalid crosshair code");
  }

  return {
    outline: bytes[4]! / 2,
    red: bytes[5]!,
    green: bytes[6]!,
    blue: bytes[7]!,
    alpha: bytes[8]!,
    splitDistance: bytes[9]!,

    innerSplitAlpha: (bytes[11]! >> 4) / 10,
    hasOutline: !!(bytes[11]! & 8),
    outerSplitAlpha: (bytes[12]! & 0xf) / 10,
    splitSizeRatio: (bytes[12]! >> 4) / 10,

    thickness: bytes[13]! / 10,
    length: bytes[15]! / 10,
    gap: signed(bytes[3]!) / 10,

    hasCenterDot: !!((bytes[14]! >> 4) & 1),
    hasAlpha: !!((bytes[14]! >> 4) & 4),
    isTStyle: !!((bytes[14]! >> 4) & 8),

    style: CrosshairStyle[
      (bytes[14]! & 0xf) >> 1
    ] as keyof typeof CrosshairStyle
  };
}
