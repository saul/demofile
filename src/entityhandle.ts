import {
  MAX_EDICT_BITS,
  NETWORKED_EHANDLE_ENT_ENTRY_MASK,
  NUM_NETWORKED_EHANDLE_BITS
} from "./consts";

export class EntityHandle {
  constructor(public readonly value: number) {}

  get isValid(): boolean {
    return this.value !== NUM_NETWORKED_EHANDLE_BITS;
  }

  get index(): number {
    return this.value & NETWORKED_EHANDLE_ENT_ENTRY_MASK;
  }

  get serialNum(): number {
    return this.value >> MAX_EDICT_BITS;
  }
}
