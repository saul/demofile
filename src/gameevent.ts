import * as assert from "assert";
import {
  CSVCMsgGameEvent,
  CSVCMsgGameEventList_descriptorT
} from "./protobufs/netmessages";

enum EventKeyType {
  TYPE_STRING = 1,
  TYPE_FLOAT,
  TYPE_LONG,
  TYPE_SHORT,
  TYPE_BYTE,
  TYPE_BOOL,
  TYPE_UINT64,
  TYPE_WSTRING
}

export class GameEvent {
  public name: string;
  public id: number;
  public keyNames: string[];

  constructor(descriptor: CSVCMsgGameEventList_descriptorT) {
    this.name = descriptor.name;
    this.id = descriptor.eventid;
    this.keyNames = descriptor.keys.map(key => key.name);
  }

  public messageToObject(
    eventMsg: CSVCMsgGameEvent
  ): Record<string, string | number | boolean | Uint8Array | Long | undefined> {
    assert(eventMsg.eventid === this.id);

    const event: Record<
      string,
      string | number | boolean | Uint8Array | Long | undefined
    > = {};
    for (let i = 0; i < this.keyNames.length; ++i) {
      const keyName = this.keyNames[i]!;
      const value = eventMsg.keys[i]!;

      switch (value.type as EventKeyType) {
        case EventKeyType.TYPE_STRING:
          event[keyName] = value.valString;
          break;
        case EventKeyType.TYPE_FLOAT:
          event[keyName] = value.valFloat;
          break;
        case EventKeyType.TYPE_LONG:
          event[keyName] = value.valLong;
          break;
        case EventKeyType.TYPE_SHORT:
          event[keyName] = value.valShort;
          break;
        case EventKeyType.TYPE_BYTE:
          event[keyName] = value.valByte;
          break;
        case EventKeyType.TYPE_BOOL:
          event[keyName] = value.valBool;
          break;
        case EventKeyType.TYPE_UINT64:
          event[keyName] = value.valUint64;
          break;
        case EventKeyType.TYPE_WSTRING:
          event[keyName] = value.valWstring;
          break;
      }
    }

    return event;
  }
}
