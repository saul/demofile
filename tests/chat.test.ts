import { DemoFile, extractPublicEncryptionKey } from "../src/index";
import { monitorProgress, startParsing, Timeline } from "./util";
import { readFile } from "fs";
import { join } from "path";

test("public matchmaking chat can be decrypted", done => {
  const demoFileName = "match730_003456990748161867873_0727052756_188.dem";
  const demo = new DemoFile();
  monitorProgress(demoFileName, demo);

  const timeline = new Timeline(demo);

  demo.userMessages.on("SayText2", um => {
    timeline.log("SayText2", `${um.msgName} ${um.params.join(" ")}`);
  });

  demo.on("end", e => {
    expect(e.error).toBeFalsy();
    timeline.assertMatches();

    done();
  });

  readFile(
    join(
      __dirname,
      "..",
      "demos",
      "match730_003456990748161867873_0727052756_188.dem.info"
    ),
    (err, data) => {
      expect(err).toBeFalsy();

      const publicKey = extractPublicEncryptionKey(data);
      demo.setEncryptionKey(publicKey);

      startParsing(demoFileName, demo);
    }
  );
});
