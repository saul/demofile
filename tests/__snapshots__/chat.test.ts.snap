// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`public matchmaking chat can be decrypted: timeline 1`] = `
Array [
  "[0] start: {\\"magic\\":\\"HL2DEMO\\",\\"protocol\\":4,\\"networkProtocol\\":13775,\\"serverName\\":\\"Valve CS:GO EU North Server (srcds106.188.41)\\",\\"clientName\\":\\"GOTV Demo\\",\\"mapName\\":\\"de_mirage\\",\\"gameDirectory\\":\\"csgo\\",\\"playbackTime\\":2843.296875,\\"playbackTicks\\":181971,\\"playbackFrames\\":90920,\\"signonLength\\":564847}",
  "[5327] SayText2: \\"Cstrike_Chat_All Black_Yuzia lol  \\"",
  "[6949] SayText2: \\"Cstrike_Chat_All lil bo peep glhf  \\"",
  "[23156] SayText2: \\"Cstrike_Chat_AllDead lil bo peep НИХУЯ ПАПОРОТНИК ТВОРИТ РЕПЧИНУ ПОД БИТ  \\"",
  "[41342] SayText2: \\"Cstrike_Chat_AllDead Black_Yuzia ns  \\"",
  "[50163] SayText2: \\"Cstrike_Chat_AllDead ?*The|Game*? =| уебак  \\"",
  "[50455] SayText2: \\"Cstrike_Chat_AllDead sss бавает  \\"",
  "[50779] SayText2: \\"Cstrike_Chat_AllDead ?*The|Game*? =| найс  \\"",
  "[51163] SayText2: \\"Cstrike_Chat_AllDead ?*The|Game*? =| та ты жуе врубвй  \\"",
  "[51331] SayText2: \\"Cstrike_Chat_AllDead ?*The|Game*? =| ты чего  \\"",
  "[51535] SayText2: \\"Cstrike_Chat_AllDead sss та я уже  \\"",
  "[51963] SayText2: \\"Cstrike_Chat_AllDead lil bo peep аоаооаоаооаоаоаооаоаоаооаоаоаоаооаоаоао  \\"",
  "[52951] SayText2: \\"Cstrike_Chat_All Папоротник Уволен  \\"",
  "[53461] SayText2: \\"Cstrike_Chat_All sss ок  \\"",
  "[54477] SayText2: \\"Cstrike_Chat_All ?*The|Game*? =| ффффффффффффффн  \\"",
  "[57997] SayText2: \\"Cstrike_Chat_AllDead sss Уволен  \\"",
  "[58575] SayText2: \\"Cstrike_Chat_AllDead Папоротник 1-1  \\"",
  "[58746] SayText2: \\"Cstrike_Chat_All sss )))  \\"",
  "[59154] SayText2: \\"Cstrike_Chat_All Папоротник Ну бот   \\"",
  "[59368] SayText2: \\"Cstrike_Chat_All Папоротник -100 к карме  \\"",
  "[59578] SayText2: \\"Cstrike_Chat_All sss зхвазпавзхва  \\"",
  "[59632] SayText2: \\"Cstrike_Chat_All sss сыглы  \\"",
  "[64868] SayText2: \\"Cstrike_Chat_All Папоротник флеш бл  \\"",
  "[65680] SayText2: \\"Cstrike_Chat_All 𝙵𝙻𝚇𝚈 niggers  \\"",
  "[92586] SayText2: \\"Cstrike_Chat_All lil bo peep HP?  \\"",
  "[93076] SayText2: \\"Cstrike_Chat_All Black_Yuzia dont' saw  \\"",
  "[100443] SayText2: \\"Cstrike_Chat_AllDead sss ты сука  \\"",
  "[100627] SayText2: \\"Cstrike_Chat_AllDead sss противная(  \\"",
  "[102011] SayText2: \\"Cstrike_Chat_AllDead sss Уволен0  \\"",
  "[102103] SayText2: \\"Cstrike_Chat_AllDead sss )))  \\"",
  "[102331] SayText2: \\"Cstrike_Chat_All lil bo peep ЗАБАНИЛИ  \\"",
  "[102411] SayText2: \\"Cstrike_Chat_All Папоротник ВСЮ ОБОЙМУ  \\"",
  "[102495] SayText2: \\"Cstrike_Chat_All lil bo peep НАХОЙ  \\"",
  "[102521] SayText2: \\"Cstrike_Chat_All Папоротник ВСЮ СУКА  \\"",
  "[102631] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[102673] SayText2: \\"Cstrike_Chat_All sss ХЗАЗВЫХЗХАВЫЗХАВЫЗХЫАВХЗАЫВЗХЫВА  \\"",
  "[102675] SayText2: \\"Cstrike_Chat_All Папоротник 13 ПУЛЬ  \\"",
  "[102683] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[102729] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[102781] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[102785] SayText2: \\"Cstrike_Chat_All Папоротник -57  \\"",
  "[102835] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[102887] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[102959] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[103017] SayText2: \\"Cstrike_Chat_All ✪sndels28 Ɑ͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞ ̶͞͞ ﻝﮞ  \\"",
  "[122140] SayText2: \\"Cstrike_Chat_All Папоротник Ну тут прям лак шо пи было у тебя)  \\"",
  "[129400] SayText2: \\"Cstrike_Chat_All lil bo peep У МЕНЯ ЕСТЬ ПИСТОЛЕТ ВОДНЫЙ ВОДНЫЙ ПИСТОЛЕТ  У МЕНЯ ЕСТЬ АВТОМАТ ВОДНЫЙ ВОДНЫЙ АВТОМАТ У МЕНЯ ЕСТЬ АППАРАТ ВОДНЫЙ ВОДНЫЙ АППАРА  \\"",
  "[137740] SayText2: \\"Cstrike_Chat_AllDead Папоротник красава  \\"",
  "[138383] SayText2: \\"Cstrike_Chat_All lil bo peep Я ЗНАЮ  \\"",
  "[138623] SayText2: \\"Cstrike_Chat_All lil bo peep Я ЛУЧШИЙ  \\"",
  "[138629] SayText2: \\"Cstrike_Chat_All Папоротник Не ты  \\"",
  "[148030] SayText2: \\"Cstrike_Chat_AllDead sss cerf  \\"",
  "[148152] SayText2: \\"Cstrike_Chat_AllDead sss СУКА  \\"",
  "[156371] SayText2: \\"Cstrike_Chat_All Папоротник тайминг штука интересная  \\"",
  "[156671] SayText2: \\"Cstrike_Chat_All sss +  \\"",
  "[156919] SayText2: \\"Cstrike_Chat_All ruf +++  \\"",
  "[156957] SayText2: \\"Cstrike_Chat_All lil bo peep +  \\"",
  "[157091] SayText2: \\"Cstrike_Chat_All sss я в своего хуярил(  \\"",
  "[163179] SayText2: \\"Cstrike_Chat_AllDead 𝙵𝙻𝚇𝚈 ????????????????????  \\"",
  "[164869] SayText2: \\"Cstrike_Chat_AllDead lil bo peep НИХУЯ ТЫ ЧИКЕН КРЫЛЫШКА  \\"",
  "[165843] SayText2: \\"Cstrike_Chat_AllDead lil bo peep ПАПОРОТНИК Я ЕБАТЬ САТИВА БРАТАН  \\"",
  "[179103] SayText2: \\"Cstrike_Chat_AllDead lil bo peep BRUH  \\"",
  "[180919] SayText2: \\"Cstrike_Chat_All ?*The|Game*? =| gg  \\"",
  "[180939] SayText2: \\"Cstrike_Chat_All Black_Yuzia gg  \\"",
  "[180943] SayText2: \\"Cstrike_Chat_All Папоротник gg  \\"",
  "[181107] SayText2: \\"Cstrike_Chat_AllDead ✪sndels28 gg u noob  \\"",
]
`;
